scipy/stats/_odds_ratio.py

```diff
@@ -0,0 +1,479 @@
+        oddsratio = np.nan
+    else:
+        oddsratio = np.inf
+    return oddsratio
+
+
+def _solve(func):
+    """
+    Solve func(nc) = 0.  func must be an increasing function.
+    """
+    # We could just as well call the variable `x` instead of `nc`, but we
+    # always call this function with functions for which nc (the noncentrality
+    # parameter) is the variable for which we are solving.
+    nc = 1.0
+    value = func(nc)
+    if value == 0:
+        return nc
+
+    # Multiplicative factor by which to increase or decrease nc when
+    # searching for a bracketing interval.
+    factor = 25.0
+    # Find a bracketing interval.
+    if value > 0:
+        nc /= factor
+        while func(nc) > 0:
+            nc /= factor
+        lo = nc
+        hi = factor*nc
+    else:
+        nc *= factor
+        while func(nc) < 0:
+            nc *= factor
+        lo = nc/factor
+        hi = nc
+
+    # lo and hi bracket the solution for nc.
+    nc = brentq(func, lo, hi, xtol=1e-13)
+    return nc
+
+
+def _nc_hypergeom_mean_inverse(x, total, ngood, nsample):
+    """
+    For the given total, ngood, and nsample, find the noncentrality
+    parameter of Fisher's noncentral hypergeometric distribution whose
+    mean is x.
+    """
+    nc = _solve(lambda nc: nchypergeom_fisher.mean(total, ngood, nsample,
+                                                   nc) - x)
+    return nc
+
+
+def _hypergeom_params_from_table(table):
+    x = table[0, 0]
+    total = table.sum()
+    ngood = table[0].sum()
+    nsample = table[:, 0].sum()
+    return x, total, ngood, nsample
+
+
+def _ci_upper(table, alpha):
+    """
+    Compute the upper end of the confidence interval.
+    """
+    if _sample_odds_ratio(table) == np.inf:
+        return np.inf
+
+    x, total, ngood, nsample = _hypergeom_params_from_table(table)
+
+    # nchypergeom_fisher.cdf is a decreasing function of nc, so we negate
+    # it in the lambda expression.
+    nc = _solve(lambda nc: -nchypergeom_fisher.cdf(x, total, ngood, nsample,
+                                                   nc) + alpha)
+    return nc
+
+
+def _ci_lower(table, alpha):
+    """
+    Compute the lower end of the confidence interval.
+    """
+    if _sample_odds_ratio(table) == 0:
+        return 0
+
+    x, total, ngood, nsample = _hypergeom_params_from_table(table)
+
+    nc = _solve(lambda nc: nchypergeom_fisher.sf(x - 1, total, ngood, nsample,
+                                                 nc) - alpha)
+    return nc
+
+
+def _conditional_oddsratio(table):
+    """
+    Conditional MLE of the odds ratio for the 2x2 contingency table.
+    """
+    x, total, ngood, nsample = _hypergeom_params_from_table(table)
+    lo, hi = hypergeom.support(total, ngood, nsample)
+
+    # Check if x is at one of the extremes of the support.  If so, we know
+    # the odds ratio is either 0 or inf.
+    if x == lo:
+        # x is at the low end of the support.
+        return 0
```

Comment:
The only way 0 could be the mean of the distribution is if the noncentrality parameter is 0 (so the random variable is identically 0). Otherwise, some of the probability mass would be greater than 0, and therefore the mean would not be 0. Correct?

Reply:
Yup.